build_tools/github_actions/repo-diff.py

```diff
@@ -0,0 +1,674 @@
+
+    # If the repo is rocm-libraries fetch from shared and projects subfolders
+    if repo == "rocm-libraries":
+        url = f"https://api.github.com/repos/ROCm/{repo}/contents/shared"
+        print(f"Requesting: {url}")
+        try:
+            data = gha_send_request(url)
+            for item in data:
+                print(f"Item: {item.get('name')} type: {item.get('type')}")
+                if item['type'] == 'dir':
+                    components.append("shared/" + item['name'])
+        except Exception as e:
+            print(f"Failed to fetch shared folder from GitHub: {e}")
+
+    # Fetch the components in the projects directory
+    url = f"https://api.github.com/repos/ROCm/{repo}/contents/projects"
+    print(f"Requesting: {url}")
+    try:
+        data = gha_send_request(url)
+        for item in data:
+            print(f"Item: {item.get('name')} type: {item.get('type')}")
+            if item['type'] == 'dir':
+                components.append("projects/" + item['name'])
+    except Exception as e:
+        print(f"Failed to fetch projects folder from GitHub: {e}")
+
+    return components
+
+def get_commits_between_shas(repo_name, start_sha, end_sha, enrich_with_files=False):
+    """
+    Get commits between two SHAs for any repository.
+
+    Args:
+        repo_name: Repository name (e.g., 'rocm-libraries', 'hip')
+        start_sha: Starting commit SHA to stop at
+        end_sha: Ending commit SHA to start from
+        enrich_with_files: If True, fetch detailed file info for each commit (needed for monorepos)
+
+    Returns:
+        List of commit objects in chronological order (newest first)
+    """
+    commits = []
+    found_start = False
+    page = 1
+    max_pages = 20
+
+    repo_type = "monorepo" if enrich_with_files else "submodule"
+    print(f"  Getting commits for {repo_type} {repo_name} from {start_sha} to {end_sha}")
+
+    while not found_start and page <= max_pages:
+        params = {"sha": end_sha, "per_page": 100, "page": page}
+        url = f"https://api.github.com/repos/ROCm/{repo_name}/commits?{urllib.parse.urlencode(params)}"
+
+        try:
+            print(f"  Fetching page {page} for {repo_name}")
+            data = gha_send_request(url)
+
+            if not data:
+                print(f"  No more commits found on page {page}")
+                break
+
+            print(f"  Page {page}: Received {len(data)} commits from API")
+
+            for commit in data:
+                sha = commit['sha']
+
+                # Enrich with file details if requested (for monorepos)
+                if enrich_with_files:
+                    print(f"  Processing commit: {sha}")
+                    try:
+                        commit_url = f"https://api.github.com/repos/ROCm/{repo_name}/commits/{sha}"
+                        commit = gha_send_request(commit_url)
+                    except Exception:
+                        pass  # Use original commit if enrichment fails
```

Comment:
This generates quite a few API requests. Is there a way to batch these into a single request?

The logs show that this is pretty fast, at least: https://github.com/ROCm/TheRock/actions/runs/19513466863/job/55858986809#step:4:125 (much faster than cloning the repository and running local git operations)